packages/ssz/src/backings/structural/array.ts

```diff
@@ -0,0 +1,180 @@
+import {ArrayLike} from "../../interface";
+import {BasicArrayType, CompositeArrayType} from "../../types";
+import {StructuralHandler} from "./abstract";
+
+export class BasicArrayStructuralHandler<T extends ArrayLike<any>> extends StructuralHandler<T> {
+  _type: BasicArrayType<T>;
+  getLength(value: T): number {
+    throw new Error("Not implemented");
+  }
+  size(value: T): number {
+    return this._type.elementType.size() * this.getLength(value);
+  }
+  assertValidValue(value: any): asserts value is T {
+    for (let i = 0; i < this.getLength(value); i++) {
+      try {
+        this._type.elementType.assertValidValue(value[i]);
+      } catch (e) {
+        throw new Error(`Invalid element ${i}: ${e.message}`);
+      }
+    }
+  }
+  equals(value1: T, value2: T): boolean {
+    for (let i = 0; i < this.getLength(value1); i++) {
+      if (!this._type.elementType.equals(value1[i], value2[i])) {
+        return false;
+      }
+    }
+    return true;
+  }
+  clone(value: T): T {
+    const newValue = this._type.structural.defaultValue();
+    for (let i = 0; i < this.getLength(value); i++) {
+      newValue[i] = this._type.elementType.clone(value[i]);
+    }
+    return newValue;
+  }
+  fromBytes(data: Uint8Array, start: number, end: number): T {
+    const elementSize = this._type.elementType.size();
+    return Array.from(
+      {length: (end - start) / elementSize},
+      (_, i) => this._type.elementType.fromBytes(data, start + (i * elementSize))
+    ) as unknown as T;
+  }
+  toBytes(value: T, output: Uint8Array, offset: number): number {
+    const length = this.getLength(value);
+    let index = offset;
+    for (let i = 0; i < length; i++) {
+      index = this._type.elementType.toBytes(value[i], output, index);
+    }
+    return index;
+  }
+  chunk(value: T, index: number): Uint8Array {
+    const output = new Uint8Array(32);
+    const itemSize = this._type.elementType.size();
+    const itemsInChunk = Math.floor(32 / itemSize);
+    const firstIndex = index * itemsInChunk;
+    // not inclusive
+    const lastIndex = Math.min(this.getLength(value), firstIndex + itemsInChunk);
+    // i = array index, grows by 1
+    // j = data offset, grows by itemSize
+    for (let i = firstIndex, j = 0; i < lastIndex; i++, j += itemSize) {
+      this._type.elementType.toBytes(value[i], output, j);
+    }
+    return output;
+  }
+}
+
+export class CompositeArrayStructuralHandler<T extends ArrayLike<any>> extends StructuralHandler<T> {
+  _type: CompositeArrayType<T>;
+  getLength(value: T): number {
+    throw new Error("Not implemented");
+  }
+  size(value: T): number {
+    if (this._type.elementType.isVariableSize()) {
+      let s = 0;
+      for (let i = 0; i < this.getLength(value); i++) {
+        s += this._type.elementType.structural.size(value[i]) + 4;
```

Comment:
I see this "4" often, it could be some constant with meaningful name :smile:

Reply:
same with 32 and 256, :+1:

Reply:
You forgot about this :disappointed: